### PR TITLE
findNumberOfRepetitions makes fewer useless arrays

### PR DESCRIPTION
**Pull request by akiezun**
Came in in profiling - findNumberOfRepetitions was creating lots of byte arrays that were discarded right away after comparing of a small range. This PR makes the comparison work on arrays directly.

@lbergelson can you have a look?

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 1]
```diff
@@ -535,17 +535,43 @@ public static int findNumberOfRepetitions(byte[] repeatUnit, byte[] testString,
...
+     *  Checks if arr1 starting at 'from' and ending right before 'to' equals to arr2.
+     *  The 'to' argument may lie outside of arr1 and then the 'elements' beyond arr1 are treated as zero.
+     *  In other words this is equivalent to:
+     <code>
+             byte[] unit = Arrays.copyOfRange(arr1, from, to);    //this allows 'to' to be outside of arr1
+             return Arrays.equals(unit,arr2);
+     </code>
+     *  but avoids making a new array instance.
+     */
+    private static boolean equalToSubarray(final byte[] arr1, final int from, final int to, final byte[] arr2){
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: This should be in `Utils` rather than embedded within this class, and should (probably) throw if `from` and `to` are out of bounds on `arr1` (unless the calling code actually requires this, which seems unlikely).

**Comment by akiezun** [comment]
how about this - i eliminated the method altogether and using Utils.equalRange

@droazen

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 2]
```diff
@@ -518,33 +518,37 @@ public static int findRepeatedSubstring(byte[] bases) {
      * @return                       Number of repetitions (0 if testString is not a concatenation of n repeatUnit's
      */
     public static int findNumberOfRepetitions(byte[] repeatUnit, byte[] testString, boolean lookForward) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Method is deprecated in GATK3 with the message: `@deprecated Move to use TandemRepeatFinder in protected (move to public if needed).`. Should we create a ticket in GATK4 for this task (migrating to `TandemRepeatFinder` here)?

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Also, since this is a public method I think you need `Utils.nonNull()` on the array args.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 3]
```diff
@@ -518,33 +518,37 @@ public static int findRepeatedSubstring(byte[] bases) {
      * @return                       Number of repetitions (0 if testString is not a concatenation of n repeatUnit's
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Add a better explanation for `lookForward` mode vs. backward mode (I was initially confused as to the intended behavior of this method in backward mode), plus an example for the backward case.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Also, `@oaram` -> `@param`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 4]
```diff
@@ -518,33 +518,37 @@ public static int findRepeatedSubstring(byte[] bases) {
...
      */
     public static int findNumberOfRepetitions(byte[] repeatUnit, byte[] testString, boolean lookForward) {
-        int numRepeats = 0;
+        final int repeatLength = repeatUnit.length;
+        final int testLength = testString.length;
+        final int lengthDifference = testLength - repeatLength;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Throw if `lengthDifference` is negative?

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: (or, if you prefer, confirm via unit test that the method does something sane in that case, and document that behavior)

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 5]
```diff
@@ -518,33 +518,37 @@ public static int findRepeatedSubstring(byte[] bases) {
...
+
         if (lookForward) {
+            int numRepeats = 0;
             // look forward on the test string
-            for (int start = 0; start < testString.length; start += repeatUnit.length) {
-                int end = start + repeatUnit.length;
-                byte[] unit = Arrays.copyOfRange(testString,start, end);
-                if(Arrays.equals(unit,repeatUnit))
+            for (int start = 0; start < testLength; start += repeatLength) {
+                if(start <= lengthDifference && Utils.equalRange(testString, start, repeatUnit, 0, repeatLength)) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: You could simplify this a bit if you made `start <= lengthDifference` the loop condition instead of `start < testLength`

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 6]
```diff
@@ -518,33 +518,37 @@ public static int findRepeatedSubstring(byte[] bases) {
...
                     break;
+                }
+            }
+            return numRepeats;
+        } else {
+            // look backward. For example, if repeatUnit = AT and testString = GATAT, number of repeat units is still 2
+            int numRepeats = 0;
+            // look backward on the test string
+            for (int start = lengthDifference; start >= 0; start -= repeatLength) {
+                if (start <= lengthDifference && Utils.equalRange(testString, start, repeatUnit, 0, repeatLength)) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: The `start <= lengthDifference` test is redundant here, since `start` is initialized to `lengthDifference` and decremented after that.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 7]
```diff
@@ -518,33 +518,37 @@ public static int findRepeatedSubstring(byte[] bases) {
...
+                if(start <= lengthDifference && Utils.equalRange(testString, start, repeatUnit, 0, repeatLength)) {
                     numRepeats++;
-                else
+                } else {
                     break;
+                }
+            }
+            return numRepeats;
+        } else {
+            // look backward. For example, if repeatUnit = AT and testString = GATAT, number of repeat units is still 2
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Move this example to the method-level javadoc

File: src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java [hunk 8]
```diff
@@ -773,6 +773,22 @@ public void testRepeatDetectorTest(RepeatDetectorTest cfg) {
...
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("T".getBytes(), "T".getBytes(), true),1);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("AT".getBytes(), "ATGATGATCATG".getBytes(), true),1);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("CCC".getBytes(), "CCCCCCCC".getBytes(), true),2);
+
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("AT".getBytes(), "GATAT".getBytes(), false), 2);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("ATG".getBytes(), "ATGATGATGATG".getBytes(), false),4);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("G".getBytes(), "ATGATGATGATG".getBytes(), false),1);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("T".getBytes(), "T".getBytes(), false),1);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("AT".getBytes(), "ATGATGATCATG".getBytes(), false),0);
+        Assert.assertEquals(GATKVariantContextUtils.findNumberOfRepetitions("CCC".getBytes(), "CCCCCCCC".getBytes(), false),2);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java]: Add test cases for both the forward and reverse cases with the length of the first byte array > the length of the second, and verify that the method returns 0.

Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java]: Also add tests for the case of null arrays.

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 9]
```diff
@@ -474,4 +474,35 @@ public void testSkimArray(final String original, final String remove) {
...
+            {"ATGATGATGATG".getBytes(), 11, 12, "G".getBytes(), true},
+            {"ATGATGATGATG".getBytes(), 10, 11, "G".getBytes(), false},
+            {"ATGATGATCATG".getBytes(), 10, 12, "AT".getBytes(), false},
+        };
+        return result;
+    }
+
+    @Test(dataProvider = "equalToSubarrayData")
+    public void testEqualToSubarray(byte[] arr1, int from, int to, byte[] arr2, boolean expected) throws Exception {
+        Assert.assertEquals(Utils.equalRange(arr1, from, arr2, 0, to - from), expected);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java]: BUGBUG: You are off by one here -- the number of positions to compare is `to - from + 1`, assuming `from` and `to` form a closed interval.

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java [hunk 10]
```diff
@@ -474,4 +474,35 @@ public void testSkimArray(final String original, final String remove) {
...
+
+            {"ATGATGATGATG".getBytes(), 11, 12, "G".getBytes(), true},
+            {"ATGATGATGATG".getBytes(), 10, 11, "G".getBytes(), false},
+            {"ATGATGATCATG".getBytes(), 10, 12, "AT".getBytes(), false},
+        };
+        return result;
+    }
+
+    @Test(dataProvider = "equalToSubarrayData")
+    public void testEqualToSubarray(byte[] arr1, int from, int to, byte[] arr2, boolean expected) throws Exception {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java]: Should be named `testEqualRange`

**Comment by droazen** [comment]
Review complete -- back to @akiezun. Merge after addressing comments.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: done https://github.com/broadinstitute/gatk/issues/1635 and fixed

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: changed language to leading vs trailing repeats

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: already documented `@return  Number of repetitions (0 if testString is not a concatenation of n repeatUnit's)`

I added tests for it

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: good point. done

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: true. fixed

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 5 above)
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Make sure you still handle the case where `lengthDifference` == `testLength` correctly, though! (ie., the `repeatLength

File: src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java (hunk 8 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtilsUnitTest.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: repeatLength == 0 can only be tolerated if testString has length 0. This is still a bogus condition so it'll disallow it (throw) and add tests for it,

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java (hunk 10 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java (hunk 9 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java]: yes, 'to' is exclusive. added a comment
